### PR TITLE
[Comb] Compare Strength Reduction

### PR DESCRIPTION
**Pull request by fyquah**
Implements comparison operation strength reduction

This is as described by the [XLS Optimizations](https://google.github.io/xls/optimizations/#comparison-operations), as referenced by https://github.com/llvm/circt/issues/1282.

This PR transforms `cmp(concat(a, b, c), concat(a, x, y)) => cmp(concat(b, c), concat(x, y))`. Some care (and tests!) are added to ensure that the sign-extension as appropriately done. This PR only handles comparing concat, which should yield some win.

Some future work could be done to recognize things like `cmp(a[31:0], {a[31:16], b[15:0]})`. I added a note about this extension.

### CODE REVIEW

**Comment by fyquah** [comment]
I decided to change my approach for this -- instead of making this part of the SimpleCannonicalizer pass, this is made into a separate pass altogether.

File: lib/Transforms/CombStrengthReduction.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace comb;
+
+#define GEN_PASS_CLASSES
+#include "circt/Transforms/Passes.h.inc"
+
+namespace {
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:32]: Please only put types in an anon namespace, static is good enough for functions and makes "local reasoning" about the code easier.

File: lib/Transforms/CombStrengthReduction.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+static std::pair<size_t, size_t> computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b)
+{
+  size_t commonPrefixLength = 0;
+  size_t commonSuffixLength = 0;
+  const size_t sizeA = a.size();
+  const size_t sizeB = b.size();
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:39]: 'const' on local variables isn't needed, plz remove

**Review by lattner — COMMENTED** [review]
High level question: why not put this into the canonicalization pass?  It seems like this monotonically simplifies the IR (always profitable), and while it is a complicated pattern match, it has a low cost filter (only applies to icmp(concat,concat)).  Putting it into canonicalization has a few advantages:
1) Each distinct pass introduces a small cost to run, so we don't want lots of micropasses.
2) Being a distinct pass means we have to decide where to run it, leading to phase ordering issues.
3) Being part of canonicalize allows the worklist to recursively simplify things after each pattern is applied.

Is there a specific reason you didn't go this way?

**Comment by fyquah** [comment]
The next step to this is attempt to capture things like `cmp(a[31:0], concat(a[31:16], b))`, which i think will start getting expensive, since it will (1) match on all `cmp` case (2) requires splitting a[31:0] internally to (a[31:16], a[15:0]).

How about this:
- Move this PR back into the cannonicalizer pass
- Implement what i described above into a separate strengthreduction pass

File: lib/Transforms/CombStrengthReduction.cpp (hunk 1 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: Done

File: lib/Transforms/CombStrengthReduction.cpp (hunk 2 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: Done

**Comment by fyquah** [comment]
@lattner  what do you think about the above suggestion?

**Comment by lattner** [comment]
I'm sorry for the delay getting back to this.  The thing to watch out for in canonicalize is "speed to match" in the negative (no match) case.   This is because there at tons and tons of patterns over time, and most of them don't trigger.  When they trigger, it typically doesn't matter how long the rewrite takes.  I'll take a look at the most recent patch, sorry again for the delay!

File: lib/Dialect/Comb/CombOps.cpp [hunk 3]
```diff
@@ -131,13 +131,26 @@ bool XorOp::isBinaryNot() {
 // ConcatOp
 //===----------------------------------------------------------------------===//
 
-void ConcatOp::build(OpBuilder &builder, OperationState &result,
-                     ValueRange inputs) {
+static unsigned valueRangeTotalWidth(ValueRange inputs)
```
Comment (lattner) [inline lib/Dialect/Comb/CombOps.cpp:134]: I'd recommend naming this `getTotalWidth`, there is no need to repeat "Value Range" as it is in the type info of the argument

File: lib/Dialect/Comb/CombOps.cpp [hunk 4]
```diff
@@ -131,13 +131,26 @@ bool XorOp::isBinaryNot() {
...
+  return resultWidth;
+}
+
+void ConcatOp::build(OpBuilder &builder, OperationState &result,
+                     ValueRange inputs) {
+  build(builder, result, builder.getIntegerType(valueRangeTotalWidth(inputs)), inputs);
+}
+
+void ConcatOp::build(OpBuilder&builder, OperationState &result, Value hd, ValueRange tl)
+{
```
Comment (lattner) [inline lib/Dialect/Comb/CombOps.cpp:149]: Plz clang format

File: lib/Dialect/Comb/CombOps.cpp [hunk 5]
```diff
@@ -131,13 +131,26 @@ bool XorOp::isBinaryNot() {
...
+  result.addOperands(ValueRange { hd });
+  result.addOperands(tl);
+  const unsigned hdWidth = hd.getType().cast<IntegerType>().getWidth();
```
Comment (lattner) [inline lib/Dialect/Comb/CombOps.cpp:152]: No need to mark local variables 'const'.

File: lib/Transforms/CombStrengthReduction.cpp [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace comb;
+
+#define GEN_PASS_CLASSES
+#include "circt/Transforms/Passes.h.inc"
+
+static std::pair<size_t, size_t> computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b)
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:32]: Pl clang format

File: lib/Transforms/CombStrengthReduction.cpp [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+      lhsOperands, rhsOperands);
+  size_t commonPrefixLength = commonPrefixSuffixLength.first;
+  size_t commonSuffixLength = commonPrefixSuffixLength.second;
+
+  auto replaceWithConstantI1 = [&](bool constant) -> LogicalResult {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(op, APInt(1, constant));
+    return success();
+  };
+
+  auto directOrCat = [&](OperandRange range) -> mlir::Value {
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:110]: Here you're passing OperandRange by value, other places it is passed by `const &`.  I'm not sure which is right, but please be consistent :)

Comment (lattner) [reply lib/Transforms/CombStrengthReduction.cpp:110]: Also, I don't think you need the `mlir::` qualifier on Value.

File: lib/Transforms/CombStrengthReduction.cpp [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+static mlir::Optional<Value> findFirstNonEmptyValue(const OperandRange &range)
+{
+  for (auto op : range) {
+    if (op.getType().getIntOrFloatBitWidth() > 0) {
+      return mlir::Optional<Value>(op);
+    }
+  }
+
+  return mlir::None;
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:81]: I don't think you need mlir:: qualifiers on None and Optional given you're including llvm.h.  Please trim them off

File: lib/Transforms/CombStrengthReduction.cpp [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+  for (auto operand : range) {
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert (width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+LogicalResult matchAndRewriteCompareConcat(ICmpOp &op, ConcatOp &lhs, ConcatOp &rhs, PatternRewriter &rewriter) {
```
Comment (lattner) [inline lib/Transforms/CombStrengthReduction.cpp:95]: Plz mark static and add a doc comment about what invariants are expected to hold when this is called.

**Review by lattner — CHANGES_REQUESTED** [review]
Taking a fresh look at this, I still think it makes sense for this transformation to be in canonicalize for all the reasons I mentioned before.  If and when there are more complicated transforms, then perhaps a dedicated pass will be justified, but these xforms shouldn't be part of it.

Thanks!

File: lib/Dialect/Comb/CombOps.cpp (hunk 3 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombOps.cpp]: done

File: lib/Dialect/Comb/CombOps.cpp (hunk 4 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombOps.cpp]: done

File: lib/Dialect/Comb/CombOps.cpp (hunk 5 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombOps.cpp]: done

File: lib/Transforms/CombStrengthReduction.cpp (hunk 7 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: I standardized everything to const OperandRange&, which i think is right. 

We definitely want it as 'const', reference is slightly less copying..

File: lib/Transforms/CombStrengthReduction.cpp (hunk 6 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: done

File: lib/Transforms/CombStrengthReduction.cpp (hunk 8 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: done

File: lib/Transforms/CombStrengthReduction.cpp (hunk 9 above)
Comment (fyquah) [reply lib/Transforms/CombStrengthReduction.cpp]: there isn't a invariant required by this function. I added a comment about what the rewriting does .

**Comment by fyquah** [comment]
Sounds good - I moved the changes into the cannonicalizer pass, and remove the strength reduction pass. thanks!

File: lib/Dialect/Comb/CombFolds.cpp [hunk 10]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
   return {};
 }
 
+// Given a range of operands, computes the number of matching prefix and
+// suffix elements. This does not perform cross-element matching.
+static std::pair<size_t, size_t>
+computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:994]: Would it be possible to implement this one direction and more generically, e.g. something like:

```
template<typename Range>
static size_t computeCommonPrefix(const Range &a, const Range &b) {
...
}
```

and then use something like:

```
auto commonPrefix = computeCommonPrefix(lhsOperands, rhsOperands);
auto commonSuffix = computeCommonPrefix(llvm::reverse(lhsOperands), llvm::reverse(rhsOperands));
```

The `llvm::reverse` algorithm is defined in STLExtras.h.  The only thing to watch for is when they are operand lists are the same, you'll get commonPrefix==commonSuffix.  In this case, just handle it as an equality compare.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 11]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+    case ICmpPredicate::sgt:
+      return replaceWithConstantI1(0);
+
+    case ICmpPredicate::sle:
+    case ICmpPredicate::sge:
+    case ICmpPredicate::ule:
+    case ICmpPredicate::uge:
+    case ICmpPredicate::eq:
+      return replaceWithConstantI1(1);
+    }
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1108]: Please call `applyCmpPredicateToEqualOperands`

File: lib/Dialect/Comb/CombFolds.cpp [hunk 12]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp
+    // are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert(width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+static Optional<Value> findFirstNonEmptyValue(const OperandRange &range) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1029]: Oh wow, nice corner case.  I'd recommend handling this a slightly different way:

1) Canonicalize cat in general to remove zero bit operands.
2) Make your new `icmp` optimization bail out with no change if it runs into the first operand being a zero bit value.

This should make it safe, catch the same cases, and will be simpler.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 13]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+  for (auto op : range) {
+    if (op.getType().getIntOrFloatBitWidth() > 0) {
+      return mlir::Optional<Value>(op);
+    }
+  }
+
+  return None;
+}
+
+static bool isPredicateSigned(ICmpPredicate predicate) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1039]: Plz move this to being a static method on ICmpOp, similar to `ICmpOp::getFlippedPredicate`

File: lib/Dialect/Comb/CombFolds.cpp [hunk 14]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+      return replaceWithConstantI1(0);
+
+    case ICmpPredicate::sle:
+    case ICmpPredicate::sge:
+    case ICmpPredicate::ule:
+    case ICmpPredicate::uge:
+    case ICmpPredicate::eq:
+      return replaceWithConstantI1(1);
+    }
+  } else {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1109]: No need for the "else" after a returning if.  Plz remove it to reduce indentation in the else block

File: lib/Dialect/Comb/CombFolds.cpp [hunk 15]
```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+    return true;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Reduce the strength icmp(concat(...), concat(...)) by doing a element-wise
+// comparison on common prefix and suffixes. Returns success() if a rewriting
+// happens.
+static LogicalResult matchAndRewriteCompareConcat(ICmpOp &op, ConcatOp &lhs,
+                                                  ConcatOp &rhs,
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1061]: Plz pass "op/lhs/rhs" all by value.  They are pointer size.

**Review by lattner — COMMENTED** [review]
This is looking good, thank you for the continued iteration on this!

File: lib/Dialect/Comb/CombFolds.cpp (hunk 15 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp:1061]: done

File: lib/Dialect/Comb/CombFolds.cpp (hunk 13 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp:1039]: Done

File: lib/Dialect/Comb/CombFolds.cpp (hunk 14 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp:1109]: Done

File: lib/Dialect/Comb/CombFolds.cpp (hunk 11 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp:1108]: Done

File: lib/Dialect/Comb/CombFolds.cpp (hunk 10 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp]: Done, i done it as follows to remove the additional corner-case.

```c++
  size_t commonPrefixLength = computeCommonPrefixLength(lhsOperands, rhsOperands);
  size_t commonSuffixLength = computeCommonPrefixLength(
      llvm::reverse(lhsOperands.drop_front(commonPrefixLength)),
      llvm::reverse(rhsOperands.drop_front(commonPrefixLength)));
```

File: lib/Dialect/Comb/CombFolds.cpp (hunk 12 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp]: It turns out that Comb.td already enforce that ConcatOp cannot take in zero-width arguments, nor can it take in zero-length varidics args. I added some defensive checks, and gotten rid of `findFirstNonEmptyValue`

**Comment by fyquah** [comment]
Most of the comments are addressed, notably, code should be somewhat simpler now with some defensive checks.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 16]
```diff
@@ -988,6 +988,155 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp
+    // are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert(width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+static bool isAnyZeroWidthOperands(OperandRange operands) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1020]: Nit but I'd rename `isAnyZeroWidthOperands` -> hasAnyZeroWidthOperands.

Also please pass OperandRange by const& for consistency

File: lib/Dialect/Comb/CombFolds.cpp [hunk 17]
```diff
@@ -988,6 +988,155 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
...
+
+  auto lhsOperands = lhs.getOperands();
+  auto rhsOperands = rhs.getOperands();
+
+  if (lhs.getType().getWidth() == 0 || rhs.getType().getWidth() == 0) {
+    return failure();
+  }
+
+  if (isAnyZeroWidthOperands(lhsOperands) ||
+      isAnyZeroWidthOperands(rhsOperands)) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1041]: Plz add some comments about why these are rejected (I know today but wont remember in a year!).  Also consider merging the two if conditions together.

Alternatively: Is the first if condition needed?  Is a zero operand concat valid?  I don't think it should be, the verifier should reject this.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 18]
```diff
@@ -1121,5 +1270,14 @@ LogicalResult ICmpOp::canonicalize(ICmpOp op, PatternRewriter &rewriter) {
     }
   }
 
+  if (auto lhs = dyn_cast_or_null<ConcatOp>(op.lhs().getDefiningOp())) {
```
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1273]: Plz add a comment describing the transformation here.

**Review by lattner — APPROVED** [review]
Looks great, please feel free to land with a few more tweaks.  Thank you for the iteration on this!

**Comment by lattner** [comment]
Actually, wait - the comb dialect doesn't permit zero width values.  I think you can completely delete all that logic.  This is checked by the concat verifier through `CombIntegerType` and `isCombIntegerType`

**Comment by fyquah** [comment]
I removed the verification code, added some notes on while verification is not needed.

File: lib/Dialect/Comb/CombFolds.cpp (hunk 18 above)
Comment (fyquah) [reply lib/Dialect/Comb/CombFolds.cpp:1274]: Done

**Comment by lattner** [comment]
Thanks, looks great, feel free to land whenever convenient

**Pull request merged**